gate-web/src/main/groovy/com/netflix/spinnaker/gate/security/userdetailsservice/google/GoogleDirectoryUserProvider.java

```diff
@@ -0,0 +1,128 @@
+import com.google.api.client.json.jackson2.JacksonFactory;
+import com.google.api.services.admin.directory.Directory;
+import com.google.api.services.admin.directory.DirectoryScopes;
+import com.google.api.services.admin.directory.model.Group;
+import com.google.api.services.admin.directory.model.Groups;
+import com.netflix.spinnaker.gate.security.AnonymousAccountsService;
+import com.netflix.spinnaker.gate.security.oauth2.client.OAuth2ClientConfig;
+import com.netflix.spinnaker.gate.security.userdetailsservice.SpinnakerUserProvider;
+import com.netflix.spinnaker.security.User;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Component;
+import org.springframework.web.client.RestTemplate;
+
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.security.GeneralSecurityException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+@Component
+@ConditionalOnProperty(value = "auth.groupMembership.service", havingValue = "google")
+public class GoogleDirectoryUserProvider implements SpinnakerUserProvider {
+
+  @Autowired
+  private GoogleDirectoryUserProviderConfig config;
+
+  @Autowired
+  private RestTemplate restTemplate;
+
+  @Autowired
+  private AnonymousAccountsService anonymousAccountsService;
+
+  @Autowired
+  private OAuth2ClientConfig oAuth2Configuration;
+
+  private static final String[] SERVICE_ACCOUNT_SCOPES = {DirectoryScopes.ADMIN_DIRECTORY_GROUP_READONLY};
+
+  private static final String[] USER_ROLES = {"user"};
+
+  private Directory getDirectoryService() throws GeneralSecurityException,
+    IOException, URISyntaxException {
+    HttpTransport httpTransport = new NetHttpTransport();
+    JacksonFactory jsonFactory = new JacksonFactory();
+    GoogleCredential credential = new GoogleCredential.Builder()
+      .setServiceAccountUser(config.getAdminUsername())
+      .setTransport(httpTransport)
+      .setJsonFactory(jsonFactory)
+      .setServiceAccountId(config.getServiceAccountEmail())
+      .setServiceAccountPrivateKeyFromP12File(new java.io.File(config.getCredentialPath()))
+      .setServiceAccountScopes(Arrays.asList(SERVICE_ACCOUNT_SCOPES))
+      .build();
+
+    return new Directory.Builder(httpTransport, jsonFactory, credential)
+      .setApplicationName("Spinnaker")
+      .build();
+  }
+
+  @Override
+  public User loadUser(String tokenValue) {
+    String userEmail = loadEmail(tokenValue);
+    Groups groups = null;
+    // TODO(jacobkiefer): refactor and deal with exceptions properly
+    try {
+      groups = loadGroups(userEmail);
+    } catch (GeneralSecurityException e) {
```

Comment:
Done. I don't think catching is the proper way to deal with the exceptions here -- maybe we can catch and subsequently throw a `UserProviderException` that would be thrown in the interface? Thoughts?